mm2src/coins/nft/trading_proto_v2/mod.rs

```diff
@@ -0,0 +1,549 @@
+            EthCoinType::Eth | EthCoinType::Erc20 { .. } => {
+                return MmError::err(ValidatePaymentError::InternalError(
+                    "EthCoinType must be Nft".to_string(),
+                ))
+            },
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn spend_nft_maker_payment_v2_impl(
+        &self,
+        args: SpendNftMakerPaymentArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let etomic_swap_contract = try_tx_s!(self.parse_contract_address(args.swap_contract_address));
+        if args.maker_secret.len() != 32 {
+            return Err(TransactionErr::Plain(ERRL!("maker_secret must be 32 bytes")));
+        }
+        let contract_type = args.contract_type;
+        let (decoded, index_bytes) =
+            try_tx_s!(self.get_decoded_tx_data_and_index_bytes(contract_type, &args.maker_payment_tx.data));
+
+        let (state, htlc_params) = try_tx_s!(
+            self.status_and_htlc_params_from_tx_data(
+                etomic_swap_contract,
+                &NFT_SWAP_CONTRACT,
+                &decoded,
+                index_bytes,
+                StateType::MakerPayments,
+            )
+            .await
+        );
+        match self.coin_type {
+            EthCoinType::Nft { .. } => {
+                let data =
+                    try_tx_s!(self.prepare_spend_nft_maker_v2_data(contract_type, &args, decoded, htlc_params, state));
+
+                self.sign_and_send_transaction(0.into(), Action::Call(etomic_swap_contract), data, U256::from(ETH_GAS))
+                    .compat()
+                    .await
+            },
+            EthCoinType::Eth | EthCoinType::Erc20 { .. } => Err(TransactionErr::ProtocolNotSupported(
+                "ETH and ERC20 Protocols are not supported for NFT Swaps".to_string(),
+            )),
+        }
+    }
+
+    pub(crate) async fn refund_nft_maker_payment_v2_timelock_impl(
+        &self,
+        args: RefundPaymentArgs<'_>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let _etomic_swap_contract = try_tx_s!(args.swap_contract_address.try_to_address());
+        let tx: UnverifiedTransaction = try_tx_s!(rlp::decode(args.payment_tx));
+        let _payment = try_tx_s!(SignedEthTx::new(tx));
+        todo!()
+    }
+
+    fn prepare_nft_maker_payment_v2_data(
+        &self,
+        contract_type: &ContractType,
+        swap_contract_address: Address,
+        token_id: U256,
+        args: &SendNftMakerPaymentArgs<'_, Self>,
+        htlc_data: Vec<u8>,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        match contract_type {
+            ContractType::Erc1155 => {
+                let function = ERC1155_CONTRACT.function("safeTransferFrom")?;
+                let amount_u256 = U256::from_dec_str(&args.amount.to_string())
+                    .map_err(|e| PrepareTxDataError::Internal(e.to_string()))?;
+                let data = function.encode_input(&[
+                    Token::Address(self.my_address),
```

Comment:
How about using `my_addr` function here instead of `self.my_address`, it will make this fix https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084#discussion_r1551838920 needed in only one place.